### PR TITLE
[2025H1] Propose a null and alignment checks goal

### PR DESCRIPTION
**Pull request by 1c3t3a**
[Rendered](https://github.com/1c3t3a/rust-project-goals/blob/ubsan-rustc-goal/src/2025h1/null-alignment-debug-checks.md)

cc: @tmandry, @RalfJung, @rcvalle

### CODE REVIEW

**Comment by RalfJung** [comment]
This mostly already exists? @saethlin added alignment checks in debug builds.

**Comment by 1c3t3a** [comment]
Oh I didn't see that! I could pivot this goal to something different in the area? E.g. around pointer arithmetics or checking enum discriminants?

File: src/2025h1/null-alignment-debug-checks.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+interacting with values that cross the language boundary (FFI, e.g. passing a
+C++ created pointer to Rust), the reasoning about such values is not always
+straightforward.
+
+At the same time, undefined behavior (UB) is triggered quickly when interacting
+with invalid pointers. E.g. [just the existence](https://lwn.net/Articles/985717/)
+of a null reference is UB, it doesn't even have to be dereferenced.
+
+Responsibility of ensuring the various invariants of the Rust
+compiler is with the programer. They have to make sure the value is not
```
Comment (programmerjake) [inline src/2025h1/null-alignment-debug-checks.md:29]:
```suggestion
compiler is with the programmer. They have to make sure the value is not
```

**Comment by 1c3t3a** [comment]
>  I would be happy to collaborate on sketching out a coherent story for our hodgepodge of runtime UB checks, but I'm a bit busy this week and next.

I would like to collaborate on this! I send you a message on Zulip to discuss this further.

For this goal: I'd love to pivot it to something helpful regarding runtime checks, so probably update it after having synced with @saethlin.

**Comment by nikomatsakis** [comment]
@1c3t3a ok, I'll hold off on merging until you and @saethlin align on the right description.

**Comment by nikomatsakis** [comment]
I haven't heard anything here, I'm going to close this PR as the initial deadline for goals has passed and I don't think it's ready to merge -- please feel free to re-open once you've got a consensus plan in place (although time IS running out...).

**Comment by 1c3t3a** [comment]
I actually pivoted the proposal here and just uploaded the new version. I was waiting for @saethlin's feedback but I think I can work that in later as well. The new PR is: #177. I hope I am not too late, sorry for the delay!

**Pull request closed without merging**